src/gleam/http.gleam

```diff
@@ -52,3 +65,170 @@ pub fn method_to_string(method) {
 
 pub external fn method_from_erlang(anything) -> Result(Method, Nil) =
   "gleam_http_native" "method_from_erlang"
+
+/// Data type containing an attribute value pair that constitute an HTTP header.
+pub type Header =
+  tuple(String, String)
+
+/// A general data type that can be be parameterised as a `Request` or `Response`
+pub type Message(head, body) {
```

Comment:
I find this indirection a little less clear than having concrete request and response types. I would be interested in looking at other similar languages and seeing how they design their data structures.

I'm also wondering if we want to be this perscriptive about HTTP request and response formats in this core language, would different servers want to use different representations? i.e. What about stateful representations like those found in Plug and Cowboy? Or those that wish to lazily parse different parts of the request?

Reply:
I think stateful representations are possible in this, because the body type is paramaterised.
For example to make something like Raxx's streaming API the body type would be.

```rust
Body{
  ContentLength(Int)
  Chunked
  None
}
```

Creating that body type would allow you to return responses where the content of the body was not part of the request, by returning chunked.

(ref https://github.com/CrowdHailer/raxx/blob/master/lib/raxx.ex#L45 in that example true meant body coming later, false meant no body, and it was also possible to return a string if you had the body already)

Reply:
Cool, thanks!

I'm still unsure about having one type abstracting over both `Request` and `Response`. Looking at the functions below it doesn't remove much duplication, and it does make conceptually a bit more complex. Is there prior art of a type combining both like this?

At the moment I would rather have two simpler types (possible in different modules so we can have simpler function names?) but perhaps there's a different name that could make it clearer? I can't think of one at present though.

Reply:
Not sure of prior art in functional languages. but when you can apply methods to request objects then obviously set_header normally exists on both.

I put a combined type because i like importing the functions into controllers and writting the following

```rust
let _ = get_header(request, "foo")
let response = set_header(response, "foo", "5")
```

If they don't work on both then you need to import get_header from the request module and set_header from the response module. And because module name and variables are both lowercase the name request/response is taken up by variables. So I am likely to end up with
```rust
let _ = request_mod.get_header(request, "foo")
let response = response_mod.set_header(response, "foo", "5")
```

Reply:
I don't feel at the moment that this is enough of a win to compensate for the increased complexity of this API. I think every HTTP lib I've used has different functions for request and response, and I don't think that has cause me any problems at least. The Message API on the other hand took me a little bit of looking to understand, and I'm very familiar with Gleam.

I think it would pay off to have this be as approachable as possible as this is probably going to be one of the first Gleam APIs that newcomers use, as people are probably going to start by trying to make a web app using Midas.

Reply:
How do you feel about opaque types? because a request/response has more constraints that the type system can manage. 
i.e. header keys are lower case and forbid certain characters. certain headers shouldn't be present multiple times, some should be combined when they are. When a body is change the content-length header should be added/updated.
I was thinking knowing the internal structure of the message type is not how you should be modifying them.

Reply:
That's a cool idea! It would mean we can't use the `.field` accessor syntax, but it would make it easier for us to iterate on the internal representation, and as you've said we could make an API that is much harder to misuse.

Reply:
For now I like this idea (of going to opaque types the most) 
Certainly the structure as it is has worked well for the applications I have written so far. 
And the more I think about the rules around content_length, (and more examples. like a 204 response not being allowed a body etc ) having users not meddle with the internals seems the right way to go.